course5/homework12/main.cc

```diff
@@ -0,0 +1,164 @@
+#include <boost/range/algorithm.hpp>
+#include <boost/program_options.hpp>
+#include <boost/algorithm/string.hpp>
+
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+#include <cstring>
+#include <string>
+#include <memory>
+#include <vector>
+#include <stdexcept>
+
+
+/***
+Homework
+
+Please, write a program that sorts text files.
+The program should be able to sort files with multiple columns/fields per line.
+The field separator, column number and sort direction should be configurable.
+Extra: There should be an option to sort by numerical value rather than text (ie “7” goes before “12”)
+
+The program should be familiar, unix-style utility:
+- The configuration should be passed by command-line options, both short (-s) and long (--long). -- help should print documentation.
+- Unless specified otherwise, the input should be consumed from stdin, and the output produced to stdout, so that the program can be used in shell pipelines
+- Any diagnostic/error messages should be printed to stderr
+Implementation requirements:
+- Use std stream API (ostream istream, fstream) for IO
+- Use Boost.ProgramOptions for handling the command-line parameters
+***/
+
+void read(std::istream& in, std::vector<std::string>& buffer)
+{
+    std::string line;
+    while (std::getline(in, line))
+    {
+        buffer.push_back(line);
+    }
+}
+
+void write(std::ostream& out, const std::vector<std::string>& buffer)
+{
+    for(auto item : buffer)
+        out << item << std::endl;
+}
+
+struct sorter
+{
+    sorter(bool descending, int column, std::string separator) :
+        m_descending{descending}
+        , m_column{column}
+        , m_separator{separator}
+        {}
+
+    bool operator() (const std::string& s1, const std::string& s2)
+    {
+        std::string to_cmp_1 = s1;
+        std::string to_cmp_2 = s2;
+
+        std::vector<std::string> split_s1;
+        std::vector<std::string> split_s2;
+
+        if (m_column != 0 && m_separator == "")
+            m_separator = " ";
+        if (m_column != 0)
+        {
+            boost::split(split_s1, s1, boost::is_any_of(m_separator));
+            boost::split(split_s2, s2, boost::is_any_of(m_separator));
+
+            to_cmp_1 = split_s1[m_column];
+            to_cmp_2 = split_s2[m_column];
+        }
+
+        if (m_descending)
+            return to_cmp_1 > to_cmp_2;
+        return to_cmp_1 < to_cmp_2;
+    }
+
+    bool m_descending = false;
+    int m_column = 0;
+    std::string m_separator = "";
+};
```

Comment:
Use lambda instead. Using functor allows you to make some complex things easily; for instance in this case you could think of several interesting performance optimizations. But since you are not doing them, just go for lambda.